src/librbd/deep_copy/ObjectCopyRequest.cc

```diff
@@ -833,6 +833,46 @@ void ObjectCopyRequest<I>::compute_dst_object_may_exist() {
                    << dendl;
 }
 
+template <typename I>
+void ObjectCopyRequest<I>::trigger_copyup() {
+  // With mirroring, a write to a non-existent primary cloned image object will 
+  // trigger a copyup which updates the older synced snaps with the parent data.
+  // This causes the rbd-mirror to ignore the parent data and only sync the new write
+  // to the clone, causing data corruption.
+  // For rbd-mirror scenarios, we trigger a copyup on the object and write the parent
+  // data to the object before writing the new clone image data.
+ldout(m_cct, 20) << "NITHYA: m_src_snap_id_start: " << m_src_snap_id_start << ", parent:" << m_src_image_ctx->parent << dendl;
+  if (m_src_snap_id_start != 0 && m_src_image_ctx->parent != nullptr) {
```

Comment:
I think you need to check `m_dst_image_ctx->parent` instead? Also, I think you should be stricter, i.e. you should try to trigger copyup only when the dst_snap_id_start is "attached" to the parent. Otherwise you will trigger copyup on every new mirrored snapshot.

Reply:
I'm not sure I understand the first point. The trigger_copyup needs to be called only on cloned images hence the check for a non-null src parent. The dst image would be a clone as well as the mirror daemon requires the parent to be mirrored in this scenario.
>> trigger copyup only when the dst_snap_id_start is "attached" to the parent.
I'll look into this.

Reply:
But why not to check the destination image for the parent (where we will need to trigger copyup)? Note, I am thinking here about a more generic scenario, not only the current rbd-mirror behavior. E.g. someone did `rbd deep-copy` with `--flatten` option (i.e. the destination image will not have a parent), and then later deep-copy only newer snapshots (like mirroring does). Currently we don't provide this feature but someone might want it for `rbd deep-copy` or some new application.

Reply:
BTW, rbd-mirror might want this `flatten` option too, e.g. when the parent image is not mirrored. My point is that you cannot assume that the destination image will always have a parent if the source image has it.

Reply:
In case of rbd mirroring, it is required. The clone will not be mirrored if the parent image is not present on the dst.

Reply:
IIUC , the flatten option with mirroring and the incremental snapshot deep copy are possible future features. We are currently not targeting mirroring without the parent images being mirrored. 
I would like  to work on the current issue where the parent is also required to be mirrored as we have a requirement to get this fixed and address the other scenarios later.